python/lsst/meas/deblender/plugins.py

```diff
@@ -30,6 +30,51 @@
 # Import C++ routines
 from .baselineUtils import BaselineUtilsF as butils
 
+
+def clipFootprintToNonzeroImpl(foot, image):
+    '''
+     Clips the given *Footprint* to the region in the *Image*
+     containing non-zero values.  The clipping drops spans that are
+     totally zero, and moves endpoints to non-zero; it does not
+     split spans that have internal zeros.
+    '''
+    x0 = image.getX0()
+    y0 = image.getY0()
+    xImMax = x0 + image.getDimensions().getX()
+    yImMax = y0 + image.getDimensions().getY()
+    newSpans = []
+    for span in foot.spans:
+        y = span.getY()
+        if y < y0 or y > yImMax:
+            continue
+        xminFound = None
+        for x in range(span.getX0(), span.getX1()+1):
+            if x < x0 or x > xImMax:
+                continue
+            if image.getArray()[y-y0, x-x0] != 0:
+                xminFound = x
+                break
+        # in no minimum was found there is no reason to look for the end, and
+        # no span will be added, i.e. the whole row is zero
+        # This is an explicit check for None, as a pixel with a value of zero
+        # will evaluate as false here
+        if xminFound is not None:
+            xmaxFound = None
+            for x in range(span.getX1(), span.getX0()-1, -1):
+                if x < x0 or x > xImMax:
+                    continue
+                if image.getArray()[y-y0, x-x0] != 0:
+                    xmaxFound = x
+                    break
+            # the max should be found, if it is not there is a problem
+            assert(xmaxFound is not None)
+            # append the new span found to the list
+            newSpans.append(afwGeom.Span(y, xminFound, xmaxFound))
+    # Time to update the SpanSet
+    foot.setSpans(afwGeom.SpanSet(newSpans))
+    foot.removeOrphanPeaks()
+
+
```

Comment:
Clipped to is a method on a SpanSet. I will evaluate if clipToNonzero should be moved to a different location, which may or may not happen as part of this ticket. I am not sure how I would feel about this being a method on a SpanSet as it really also only makes sense in the context of a specific image and not something geometric, but I could be persuaded one way or the other.
In relation to your speed question, I tweaked the routine to be even faster than the old c++ equivalent.